### PR TITLE
Deque

### PR DESCRIPTION
**Pull request by Casey0Kane**

### CODE REVIEW

File: src/deque.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    def pop(self):
+        """Remove a value from the end of the deque and returns it
+        (raises an exception if the deque is empty)."""
+        if self.head is None:
+            raise ValueError("Cannot pop from an empty deque.")
+        val = self.head.val
+        self.head = self.head.next_node
+        self.head.prev_node = None
+        self.persist = False
+        return val
```
Comment (bgarnaat) [inline src/deque.py:62]: This probably shouldn't happen... 
```python
In [2]: dq = Deque()

In [3]: dq.append(1)

In [4]: dq.pop()
---------------------------------------------------------------------------
AttributeError: 'NoneType' object has no attribute 'prev_node'
```
or this ...
```python
In [23]: dq = Deque()

In [24]: dq.appendleft(1)

In [25]: dq.pop()
---------------------------------------------------------------------------
AttributeError: 'NoneType' object has no attribute 'next_node'
```

File: src/deque.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    def popleft(self):
+        """Remove a value from the front of the deque and returns it
+        (raises an exception if deque is empty)."""
+        if self.tail is None:
+            raise ValueError("Cannot popleft from an empty deque.")
+        val = self.tail.val
+        self.tail = self.tail.prev_node
+        self.tail.next_node = None
+        self.persist = False
+        return val
```
Comment (bgarnaat) [inline src/deque.py:73]: This probably shouldn't happen either
```python
In [12]: dq = Deque()

In [13]: dq.append(1)

In [14]: dq.popleft()
---------------------------------------------------------------------------
AttributeError: 'NoneType' object has no attribute 'next_node'
```

or this... 
```python
In [29]: dq = Deque()

In [30]: dq.appendleft(1)

In [31]: dq.popleft()
---------------------------------------------------------------------------
AttributeError: 'NoneType' object has no attribute 'next_node'
```

File: src/deque.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        if not self.persist:
+            self.length = 0
+            if self.head:
+                self.length = 1
+            curr = self.head
+            while curr is not self.tail:
+                curr = curr.next_node
+                self.length += 1
+            return self.length
+        return self.length
```
Comment (bgarnaat) [inline src/deque.py:100]: Think about how to make size() an O(1) operation.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+
+setup(
+    name="echo",
+    description="A project to build a basic echo server.",
+    version=0.1,
+    author="Patrick & Joey & Avery",
+    author_email="",
+    license="MIT",
+    py_modules=['linked_list', 'stack'],
```
Comment (bgarnaat) [inline setup.py:14]: Think about updating your list of py_modules to reflect the work you've done.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+"""Setup file for data structures assignment."""
+
+
+from setuptools import setup
+
+
+setup(
+    name="echo",
+    description="A project to build a basic echo server.",
```
Comment (bgarnaat) [inline setup.py:9]: Not sure the deque (or other data structures) work as an echo server

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    author_email="",
+    license="MIT",
+    py_modules=['linked_list', 'stack'],
+    package_dir={'': 'src'},
+    install_requires=[''],
+    extras_require={'test': ['pytest', 'pytest-watch', 'pytest-cov', 'tox']},
+    entry_points={
+        # 'console_scripts': [
+        #     "command = module_name:main",
+        # ]
```
Comment (bgarnaat) [inline setup.py:21]: Less corpse-code!

File: src/deque.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    peekleft(): returns the next value that would be returned by popleft but leaves the value in the
+    deque (returns None if the deque is empty).
+
+    size(): returns the count of items in the queue (returns 0 if the queue is empty)."""
+    def __init__(self):
+        """Create a new instance of a doubly-ended queue."""
+        self.head = None
+        self.tail = None
+        self.length = 0
+        self.persist = True
```
Comment (nhuntwalker) [inline src/deque.py:29]: What is this persist flag for?

**Pull request merged**